### PR TITLE
Enums instead of strings

### PR DESCRIPTION
**Pull request by exocom**
## All Submissions:
-------------------------------------
* [YES] Have you added an explanation of what your changes do and why you'd like us to include them?
* [NA] I have updated the documentation accordingly.
* [NA] I have added tests to cover my changes.
* [YES] All new and existing tests passed.
* [YES] My code follows the code style of this project.
* [YES] I ran lint checks locally prior to submission.
* [YES] Have you checked to ensure there aren't other open PRs for the same update/change?

## What is the current behavior?
-------------------------------------

When there is a list of strings the string values are hard coded. This makes updating and refactoring difficult.

Issue Number: N/A

## What is the new behavior?
-------------------------------------

Use enums (struct of string) for strings. Many of the strings are already defined in the Azure DevOps sdk. Also used simplified casting instead of writing expand methods.

- `schema.ValidateFunc` of string list will be enums instead of strings. Formatted multi line for consistency.
- `schema.Default` when enum exists will use enum instead of string
- When enum doesn't exist one is created and placed at the top of most relevant file.
- When flattening value simply pass string into enum for conversion. `visibility := core.ProjectVisibility(d.Get("visibility").(string))`
- Cleaned up error messages.
- Rename func flattenReleaseDefinitionTriggers TO flattenBuildDefinitionTriggers

## Does this introduce a breaking change?
-------------------------------------
- [NO]

## Any relevant logs, error output, etc?
-------------------------------------
(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)

## Other information
-------------------------------------

### CODE REVIEW

**Comment by tmeckel** [comment]
@exocom YAYY!! 😁

File: azuredevops/resource_build_definition.go [hunk 1]
```diff
@@ -701,9 +718,9 @@ func expandBuildDefinition(d *schema.ResourceData) (*build.BuildDefinition, stri
 	repository := repositories[0].(map[string]interface{})
 
 	repoName := repository["repo_name"].(string)
-	repoType := repository["repo_type"].(string)
+	repoType := RepoType(repository["repo_type"].(string))
 	repoURL := ""
-	if strings.EqualFold(repoType, "github") {
+	if repoType == RepoTypeValues.GitHub {
```
Comment (nmiodice) [inline azuredevops/resource_build_definition.go:723]: Can we revert this back to a case-insensitive comparison?

Comment (exocom) [reply azuredevops/resource_build_definition.go:735]: @nmiodice Thanks for the quick review. I've merged in the Bitbucket stuff.

The enums are setup as Case Sensitive. The terraform validation will fail if the values are not entered exactly. This means we don't have to check Insensitively. Yahoo!

This is done by always setting ignoreCase to false in the `validation.StringInSlice`. Everywhere a list was setup this was already in place.

```
ValidateFunc: validation.StringInSlice([]string{... false),
```

image

Comment (tmeckel) [reply azuredevops/resource_build_definition.go:735]: Well I wouldn't annoy people with such nitty gritty stuff if it's not really required. Personally I like systems that allow me to enter string the way I like if it's not really required to follow any rules like capitalization and stuff. Why are the validation function is setup like this. I mean `ignoreCase = false`. Any special reason that stems from the DevOps REST API? Or was it just implemented like that?

Comment (exocom) [reply azuredevops/resource_build_definition.go]: @tmeckel Yes, the ignore case was previously implemented like that. So just following the coding style.

However, I did a quick check on the API and it doesn't care about case. IE it is case Insensitive. So I went thru and update all of the string comparisons to use strings.EqualFold. Also found a couple of places where "bitbucket" was a string. So I switched it to using the enum.

@nmiodice This PR is now good to go! 

Thanks you both for the feedback. Love working on this project.

Comment (exocom) [reply azuredevops/resource_build_definition.go]: Reverted back to case-insensitive comparison.

Comment (tmeckel) [reply azuredevops/resource_build_definition.go]: Great job!! Really appreciate that you did some analysis and reverted back to case-insensitive mode!

**Comment by nmiodice** [comment]
Great work!!

**Comment by exocom** [comment]
@nmiodice @tmeckel I was wrong about the case sensitive. The API doesn't care, but the actual system code does seem to care. We might want to consider going case-sensitive after all.

Turns out there are **parts** of the API payloads that are case sensitive.  On my branch I just sent an enum as "build" instead of "Build", because I used the SDKs enum, this caused a runtime exception. Changing it back to "Build" fixed it. 

Oddly the API doesn't check case and then you end up with a runtime exception 🙀

I can see solving this in a few ways. 

1 could normalize case sensitive enums during expand to always be Title case when sent to the APi.
2. could force user to enter correct case. IE make them case sensitive through out.
3. could report this as a bug and see if the team can make the code side of the system insensitive.


What are your guys' thoughts? Any other ideas?

<br><br>



json

image

error-side

error

**Comment by nmiodice** [comment]
(2) seems to be the easiest to maintain

**Pull request merged**